ipn/ipnlocal/local.go

```diff
@@ -3522,18 +3523,99 @@ func (b *LocalBackend) setTCPPortsInterceptedFromNetmapAndPrefsLocked() {
 				}
 			}
 			if b.serveConfig.Valid() {
+				servePorts := make([]uint16, 0, 3)
 				b.serveConfig.TCP().Range(func(port uint16, _ ipn.TCPPortHandlerView) bool {
 					if port > 0 {
-						handlePorts = append(handlePorts, uint16(port))
+						servePorts = append(servePorts, uint16(port))
 					}
 					return true
 				})
+				handlePorts = append(handlePorts, servePorts...)
+				// don't listen on netmap addresses if we're in userspace mode
+				if !wgengine.IsNetstack(b.e) {
+					b.updateServeTCPPortNetMapAddrListenersLocked(servePorts)
+				}
 			}
 		}
 	}
 	b.setTCPPortsIntercepted(handlePorts)
 }
 
+// updateServeTCPPortNetMapAddrListenersLocked starts a net.Listen for configured
+// Serve ports on all the node's addresses.
+// Existing Listeners are closed if port no longer in incoming ports list.
+//
+// b.mu must be held.
+func (b *LocalBackend) updateServeTCPPortNetMapAddrListenersLocked(ports []uint16) {
+	nm := b.netMap
+	if nm == nil {
+		b.logf("netMap is nil")
+		return
+	}
+	if nm.SelfNode == nil {
+		b.logf("netMap SelfNode is nil")
+		return
+	}
+
+	if b.serveListeners == nil {
+		b.serveListeners = make(map[netip.AddrPort]net.Listener)
+	}
+
+	for ap, ln := range b.serveListeners {
+		if !slices.Contains(ports, ap.Port()) {
+			b.logf("closing listener %v", ap)
+			ln.Close()
+			delete(b.serveListeners, ap)
+		}
+	}
+
+	for _, a := range nm.Addresses {
+		for _, p := range ports {
+			addrPort := netip.AddrPortFrom(a.Addr(), p)
+			network := "tcp"
+			if addrPort.Addr().Is6() {
+				network = "tcp6"
+			}
+
+			if _, ok := b.serveListeners[addrPort]; ok {
+				continue // already listening
+			}
+
+			ln, err := net.Listen(network, addrPort.String())
+			if err != nil {
+				b.logf("error listening, %s on %v: %v", network, addrPort, err)
+				return
+			}
+
+			b.serveListeners[addrPort] = ln
+
+			b.logf("listening %s on %s", network, addrPort)
+			go b.handleServeListenersAccept(ln, p)
+		}
+	}
+}
+
+// handleServeListenersAccept accepts connections for the serveListeners.
+// Calls HandleInterceptedTCPConn in a new goroutine for each accepted connection.
+func (b *LocalBackend) handleServeListenersAccept(ln net.Listener, port uint16) {
+	defer ln.Close()
+
+	for {
+		conn, err := ln.Accept()
+		if err != nil {
+			b.logf("error accepting connection: %v", err)
```

Comment:
if this was an intentional shutdown we shouldn't log a scary "listener closed EOF in use blah blah" message

Maybe each listener should be a type with a context and knowing whether it wants to be running or not. Then that listener can have a Close method that sets an atomic.Bool (or closes a context) and then closes the listener.

Then we can actually move the net.Listen into the serveListener.run goroutine and on listen failure (windows is weird) it can retry with a backoff (see the "backoff" type, which can backoff with a context, the same one you'd close on serveListener.close)

Then, if you get a real error from ln.Accept, you revert back to doing a net.Listen again (outer loop) until your context is closed.